index.bs

```diff
@@ -1797,6 +1822,48 @@ a [=source type=] |sourceType|, and a [=duration=] |expiry|:
     1. Set |startDuration| to |endDuration|.
 1. Return |windows|.
 
+To <dfn>parse trigger specs</dfn> given a [=map=] |map|, a [=moment=]
+|sourceTime|, a [=duration=] |expiry|, a [=report window list=]
+|topLevelWindows|, and an unsigned 32-bit integer |defaultTriggerDataCardinality|:
+
+1. [=Assert=]: |defaultTriggerDataCardinality| is greater than 0 and less than
+    or equal to [=max distinct trigger data per source=].
+1. Let |specs| be a new [=map=].
+1. If |map|["`trigger_specs`"] does not [=map/exist=]:
+    1. Let |spec| be a new [=trigger spec=] with the following items:
+        : [=trigger spec/event-level report windows=]
+        :: |topLevelWindows|
+    1. For each integer |triggerData| between 0 (inclusive) and
+        |defaultTriggerDataCardinality| (exclusive):
+        1. [=map/Set=] |specs|[|triggerData|] to |spec|.
+    1. Return |specs|.
+1. If |map|["`trigger_specs`"] is not a [=list=] or [=list/is empty=], or its
+    [=list/size=] is greater than [=max distinct trigger data per source=],
+    return an error.
+1. [=list/iterate|For each=] |item| of |map|["`trigger_specs`"]:
+    1. If |item| is not a [=map=], return an error.
+    1. Let |spec| be a new [=trigger spec=] with the following items:
+        : [=trigger spec/event-level report windows=]
+        :: |topLevelWindows|
+    1. If |item|["`event_report_windows`"] [=map/exists=]:
+        1. Let |reportWindows| be the result of
+            [=parsing report windows ignoring legacy=] with |map|, |sourceTime|,
+            and |expiry|.
+        1. If |reportWindows| is an error, return it.
+        1. Set |spec|'s [=trigger spec/event-level report windows=] to
+            |reportWindows|.
+    1. If |item|["`trigger_data`"] does not [=map/exist=], is not a [=list=], or
+        [=list/is empty=], or its [=list/size=] is greater than
```

Comment:
Add the comment here as well on whether we can allow the no-op configuration for an empty list of "trigger_data".

Reply:
There's no reason to specify an empty list. I would rather make it an error.

That said, we *could* make it so that the `trigger_data` field itself is optional, and if it's omitted it could default to the sequence `[0, defaultTriggerDataCardinality)` which could allow optimization in some cases in which the reporting origin wants to use e.g. a non-default `summary_window_operator` but the default trigger data values.

Reply:
I don't feel strongly, but as empty trigger_specs is now allowed, this may be the only empty list that will be rejected. Allowing empty trigger_data is no-op, but makes the size checking on the top level difficult, so probably it's fine leaving as is.

Reply:
BTW, I feel it's a bit similar to source_keys field in aggregatable_trigger_data, which allows the no-op empty list.

Reply:
`event_report_windows.end_times` [also cannot be empty](https://wicg.github.io/attribution-reporting-api/#parse-report-windows).

The yet-to-be-specified `summary_buckets` also cannot be empty.

Reply:
That's fair,thanks.